### ISSUE TITLE
[Web] App crashes on safari on iOS when loading large images

### ISSUE THREAD

**Issue by polina-c** [flutter/flutter#152709]
This reproduces reliably on all iPhones.

To repro:
1. Open https://vitalsynergyoregon.web.app in iPhone
2. Observe how the website opens couple times and then error page (below in comment)

- source code: https://github.com/vitalsynergy/vital_synergy
- deployed website: https://vitalsynergyoregon.web.app/
- i deployed this app by running `firebase deploy` in the folder `website`. I tried both Flutter master and Flutter stable, also i tried to recreate firebase project and deploy to it.

what SDKs I tried:
```
flutter --version
Flutter 3.24.0-1.0.pre.214 • channel master • git@github.com:flutter/flutter.git
Framework • revision c63733310f (9 days ago) • 2024-07-20 19:26:30 -0400
Engine • revision 207f088f73
Tools • Dart 3.6.0 (build [3.6.0-55.0.dev](http://3.6.0-55.0.dev/)) • DevTools 2.37.1
```

```
 flutter --version
Flutter 3.22.3 • channel stable • git@github.com:flutter/flutter.git
Framework • revision b0850beeb2 (13 days ago) • 2024-07-16 21:43:41 -0700
Engine • revision 235db911ba
Tools • Dart 3.4.4 • DevTools 2.34.3
```

**Comment by polina-c** [comment]
![image](https://github.com/user-attachments/assets/efeb0328-67a0-4859-bcf6-335cde057d02)

**Comment by danagbemava-nc** [comment]
Reproducible using the code sample in the repo provided above. 

I managed to reduce the code sample to what was actually causing the crash. The issue is caused by trying to load 
https://github.com/vitalsynergy/vital_synergy/blob/main/website/assets/images/logo_large.jpg

The reduced code sample is posted below. The image is about 18.7mb so I can't upload it in the comment (GH  apparently has a 10mb image limit)

As noted above, this works fine on android and desktop browsers.  

<details>
<summary>code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: Scaffold(
        body: Center(child: Image.asset('assets/logo_large.jpg',),),
      ),
    );
  }
}
```
</details>

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 3.22.3, on macOS 14.5 23F79 darwin-arm64, locale en-GB)
    • Flutter version 3.22.3 on channel stable at /Users/nexus/dev/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision b0850beeb2 (2 weeks ago), 2024-07-16 21:43:41 -0700
    • Engine revision 235db911ba
    • Dart version 3.4.4
    • DevTools version 2.34.3
```


```
[!] Flutter (Channel master, 3.24.0-1.0.pre.410, on macOS 14.5 23F79 darwin-arm64, locale en-GB)
    • Flutter version 3.24.0-1.0.pre.410 on channel master at /Users/nexus/dev/sdks/flutters
    ! Warning: `flutter` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision dc99d2f692 (6 hours ago), 2024-08-01 21:40:10 -0400
    • Engine revision fe39ed980e
    • Dart version 3.6.0 (build 3.6.0-107.0.dev)
    • DevTools version 2.38.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.
```
</details>

**Comment by polina-c** [comment]
Reduced size and, yes, it works. Thank you!

**Comment by maharshi-sinha** [comment]
I had faced the same error once; this might be caused by a few things: 

- High Memory Usage: Large images consume a lot of memory, which can lead to app crashes on devices with limited resources. iOS is particularly sensitive to high memory usage, and the system may terminate your app if it exceeds available memory.

- Rendering Overhead: Displaying large images requires significant processing power, which can cause slow rendering, leading to jank (lag or stutter) or crashes.

- Network Load: If the images are loaded from the network, downloading large files can lead to timeouts or slow loading times, affecting the user experience.

- Browser Limitations: When using Safari or other browsers on iOS, there may be additional constraints on memory and processing that exacerbate these issues.

I'll suggest either you can compress the size of the image or your asset or there's one package [cached_network_image](https://pub.dev/packages/cached_network_image) for the loading & caching of images efficiently

**Comment by caseyryan** [comment]
Any updates / ideas about this? The issue is very annoyoing on ios

**Comment by kevmoo** [comment]
Hey @caseyryan – can you confirm the version of iOS and the hardware you're testing on?
Also, would you provide a full code sample we can reproduce with?

**Comment by caseyryan** [comment]
> Hey [@caseyryan](https://github.com/caseyryan) – can you confirm the version of iOS and the hardware you're testing on? Also, would you provide a full code sample we can reproduce with?

Hey. It's iOS versions starting from 15 up to the latest, it doesn't matter, iPhone 13 mini. The app also doesn't matter, any long  list with network images will force the reload, even if the image loading has stopped. But as I've already found out yesterday, the problem is not in Flutter, safari will crash any web app like this even some React app, including Instagram, so I don't think anything can be done on Flutter side. Btw, google chrome works with the same app on the same device flawlessly, even though it's also base on webkit (afaik). I have no idea how to explain this

**Comment by yjbanov** [comment]
We should test this again. Last year we changed how images are decoded and uploaded to the GPU in Safari (https://github.com/flutter/engine/pull/53201), which should have helped a lot here. If this is still an issue, we should into what else might be going on here.

**Comment by caseyryan** [comment]
> We should test this again. Last year we changed how images are decoded and uploaded to the GPU in Safari ([flutter/engine#53201](https://github.com/flutter/engine/pull/53201)), which should have helped a lot here. If this is still an issue, we should into what else might be going on here.

Well, unfortunately, the problem is still present. Flutter 3.29.2 stable. Yesterday, I've installed the newest iOS, and checked it again. For a moment i was almost happy because it looked like the problem is gone. But after i few more tests it crashed again 😭. On android this problem is not present at all but there is another, the scroll of the list with images is not perfectly smooth, even with very small ones (like 6-8 kb). If i replace the images with simple SizedBox or a Container, then it works perfectly. Btw, except the crash, everything else on iOS is perfect. The scroll is very smooth even if there's a lot of images

**Comment by nikhil1198** [comment]
From my last experiments with Image.network on flutter web, it seems that on ios browsers the setting - WebHtmlElementStrategy, is the culprit. Even with a long list of images the WebHtmlElementStrategy.never works well, but with WebHtmlElementStrategy.prefer or WebHtmlElementStrategy.fallback, the browser reloads and then crashes.

For now I have handled it using:
```
WebHtmlElementStrategy webHtmlElementStrategy = isiOSBrowser
      ? WebHtmlElementStrategy.never
      : WebHtmlElementStrategy.prefer;
```

But this leads to CORS images not loading which need to be then handled separately.

Need to fix the html rendering of the images in flutter web on ios browsers in Image.network()/NetworkImage()

**Comment by nikhil1198** [comment]
> > From my last experiments with Image.network on flutter web, it seems that on ios browsers the setting - WebHtmlElementStrategy, is the culprit. Even with a long list of images the WebHtmlElementStrategy.never works well, but with WebHtmlElementStrategy.prefer or WebHtmlElementStrategy.fallback, the browser reloads and then crashes.
> > For now I have handled it using:
> > ```
> > WebHtmlElementStrategy webHtmlElementStrategy = isiOSBrowser
> >       ? WebHtmlElementStrategy.never
> >       : WebHtmlElementStrategy.prefer;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But this leads to CORS images not loading which need to be then handled separately.
> > Need to fix the html rendering of the images in flutter web on ios browsers in Image.network()/NetworkImage()
> 
> It worked only with HTML renderer. It doens't work with canvaskit. I tested all of them, it doesn't affect anything now

Actually with Flutter 3.29.2, which I am using, there is no rendering with HTML available build building flutter web. I have to therefore select the image rendering from Image.network(). In the last version of Flutter I used to build with --web-renderer, then the network images used to work without any issues. Are you building the flutter web with HTML on previous version of Flutter or are you able to make it work on the latest version of Flutter with Image.network()?

**Comment by caseyryan** [comment]
> > > From my last experiments with Image.network on flutter web, it seems that on ios browsers the setting - WebHtmlElementStrategy, is the culprit. Even with a long list of images the WebHtmlElementStrategy.never works well, but with WebHtmlElementStrategy.prefer or WebHtmlElementStrategy.fallback, the browser reloads and then crashes.
> > > For now I have handled it using:
> > > ```
> > > WebHtmlElementStrategy webHtmlElementStrategy = isiOSBrowser
> > >       ? WebHtmlElementStrategy.never
> > >       : WebHtmlElementStrategy.prefer;
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > But this leads to CORS images not loading which need to be then handled separately.
> > > Need to fix the html rendering of the images in flutter web on ios browsers in Image.network()/NetworkImage()
> > 
> > 
> > It worked only with HTML renderer. It doens't work with canvaskit. I tested all of them, it doesn't affect anything now
> 
> Actually with Flutter 3.29.2, which I am using, there is no rendering with HTML available while building flutter web. I have to therefore select the image rendering from Image.network(). In the last version of Flutter I used to build with --web-renderer, then the network images used to work without any issues. Are you building the flutter web with HTML on previous version of Flutter or are you able to make it work on the latest version of Flutter with Image.network()?

No, I'm using canvas kit with the latest version of Flutter 3.29.2, on iOS. The problem is still there, regardless of the selected renderer. I even updated the iOS itself to the latest version. It works a bit longer without a crash there (like maybe 2 times as long) but in the end it still crashes

**Comment by sangaritta** [comment]
I'm having the exact same issue, but my app will crash on any Safari device, no matter the device, (I'm using cached_images) and just one 3000x3000 image will make the whole app crash and go in a error loop until safari just stops the page, also happens on Brave Browser.

I'm going to try using small image previews, but still I second the importance of this error

**Comment by sangaritta** [comment]
This only happens on iOS, Android, Windows, works well

**Comment by caseyryan** [comment]
> This only happens on iOS, Android, Windows, works well

Do you have any problems with scroll smoothness on android? 
There's some strange stuff happening too. For example, Chrome works absolutely perfect, the scroll is smooth and fast but in any webview it apparently works much worse. I wouldn't say very bad but compared to Chrome it's obviously worse. Maybe chrome uses some very advanced render engine, idk, but the fact it that on;y in chrome the same app works very good

**Comment by sangaritta** [comment]
https://github.com/user-attachments/assets/fe483bd9-fa1f-4626-a285-a6c166f1375d

yes, the scroll seems to be laggy on Android WEB when there are network images, I was able to recreate the issue

**Comment by pastelcode** [comment]
Bug present in 3.32.4 ;(
Deployed web site: [www.lampo.design](https://www.lampo.design/#/shop/3s44gro0u41wpm7), in product details the view crashes due to several images list.
I reduced the images size as much as possible (images are around 80 kb each one) but bug still happens.

**Comment by yjbanov** [comment]
@pastelcode It seems the images are heavily _compressed_ but their _size_ is still big. For example, [this image](https://api.lampo.design/api/files/pbc_2415318885/g53l9dr6os3vs15/img_5431_zjoj0jmule.webp) is 92kb compressed, but it's 4032 x 3024 pixels in size (~50 MB). Flutter has to decompress the image before rendering it to screen, so the compressed size won't affect memory usage. If you load 20 images like that, you'd need 1GB of GPU memory to hold them (and depending on the operations done with the image, sometimes CPU memory as well). You will notice that even in Chrome, where we have pretty good image support, there's jank because these images need to be pushed to WebGL via `textSubImage2d` and that causes jank, even if it's not crashing the browser.

Flutter could try downscaling images automatically for you, but unfortunately it currently doesn't, and it would be a non-trivial project to make it work (BTW, this could be a useful pub package - and fun to build - if someone is interested in tackling it).

**Comment by sangaritta** [comment]
Can we please raise awareness on this issue? It's been present for many versions and there is still no intent to fix it. My best bet would it be to lazy load the images

I managed to fix this issue by keeping 2 versions of the image in my object storage (Original & Downscaled). But this requires more processing power on the backend which is always not ideal.

**Comment by lutes1** [comment]
I am experiencing this same issue when rendering web pages on top of my flutter widget tree in an `HTMLIFrameElement`. Works perfectly on `Android/Desktop` but crashes regularly on an `iPhone`, regardless of the browser. It looks like an OOM issue as I've been able to regularly catch OOM event using Xcode Instruments. The website that I'm rendering in the iframe is not particularly super heavy, it renders a simple 3d scene with `threejs`. Let me know I can help with further details.

**Comment by PrashanthKumarDS-yourdost** [comment]
WebHtmlElementStrategy webHtmlElementStrategy = isiOSBrowser
      ? WebHtmlElementStrategy.never
      : WebHtmlElementStrategy.prefer; this working, no crash. But images are not coming

**Comment by jeedy20** [comment]
I’m facing the same issue. Is there any fix for this? It’s urgent for me.

**Comment by sangaritta** [comment]
@jeedy20 Keep a downscaled version of the image(s) on your server, not ideal but it was the only solution for me

**Comment by jeedy20** [comment]
> [@jeedy20](https://github.com/jeedy20) Keep a downscaled version of the image(s) on your server, not ideal but it was the only solution for me

But i didn't experience this issue with html renderer. I guess i have to downgrade my flutter version.

**Comment by jiahaog** [comment]
Just by observation of some crashing cases I don't think it's a memory leak. It seems like the issue has to do with peak memory where Safari decides to crash the page should that reach a limit.

This is observed on on a physical iPhone 13 mini (iOS 18).

```
$ flutter --version
Flutter 3.39.0-1.0.pre-138 • channel master • git@github.com:jiahaog/flutter
Framework • revision f954fb79dd (6 hours ago) • 2025-11-16 11:51:26 -0500
Engine • hash 65e5cbf2f725c6ae2775382d44027332c9072edb (revision f954fb79dd) (5 hours ago) • 2025-11-16 16:51:26.000Z
Tools • Dart 3.11.0 (build 3.11.0-142.0.dev) • DevTools 2.52.0
```

### Repro

It loads 20 large images at the same time on startup, which causes the page to crash immediately:

<details>
  <summary>Repro Code</summary>
  
```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MaterialApp(home: Scaffold(body: SimpleImageList())));
}

class SimpleImageList extends StatelessWidget {
  final imageUrls = List.generate(
    20,
    (i) => 'https://picsum.photos/seed/$i/4000',
  );

  SimpleImageList({super.key});

  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      child: Column(
        children: [
          for (final path in imageUrls)
            SizedBox(width: double.infinity, child: _buildImage(path)),
        ],
      ),
    );
  }

  Widget _buildImage(String path) =>
      Image.network(path, fit: BoxFit.cover, height: 120.0);
}
```
</details>

### Workaround

The following applies a workaround to the repro. Googlers, see internal bug b/458880675 for details. Both of the following are needed:

- Load images one at a time (or tune the argument to [`Pool`](https://pub.dev/documentation/pool/latest/pool/Pool/Pool.html)) to trade latency for not crashing
- Apply `cacheWidth` to reduce the size of the image that is cached by Flutter (see ["Memory Usage"](https://api.flutter.dev/flutter/widgets/Image-class.html))


<details>
  <summary>Repro with Workaround</summary>

```dart
import 'dart:async';
import 'package:pool/pool.dart';

import 'package:flutter/material.dart';

void main() {
  runApp(MaterialApp(home: Scaffold(body: SimpleImageList())));
}

class SimpleImageList extends StatelessWidget {
  final imageUrls = List.generate(
    20,
    (i) => 'https://picsum.photos/seed/$i/4000',
  );

  SimpleImageList({super.key});

  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      child: Column(
        children: [
          for (final path in imageUrls)
            SizedBox(width: double.infinity, child: _buildImage(path)),
        ],
      ),
    );
  }

  Widget _buildImage(String path) {
    final networkImage = NetworkImage(path);
    // To override the image provider of `Image.network`, we need to duplicate some
    // portions of the constructor here.
    return Image(
      image: ResizeImage.resizeIfNeeded(
        // Reduce the size of the iamge that is cached by Flutter.
        /*cacheWidth=*/ 1000,
        null,
        SequentialNetworkImage(delegate: networkImage),
      ),
      fit: BoxFit.cover,
      height: 120.0,
    );
  }
}

class SequentialNetworkImage extends ImageProvider<NetworkImage> {
  final NetworkImage delegate;

  SequentialNetworkImage({required this.delegate});

  @override
  Future<NetworkImage> obtainKey(ImageConfiguration configuration) =>
      delegate.obtainKey(configuration);

  @override
  ImageStreamCompleter loadImage(
    NetworkImage key,
    ImageDecoderCallback decode,
  ) => _SequentialImageStreamCompleter(delegate, key, decode);
}

class _SequentialImageStreamCompleter extends ImageStreamCompleter {
  /// Global executor ensures that only one image can be loaded at a time.
  ///
  /// https://pub.dev/packages/pool
  static final _executor = Pool(1);

  _SequentialImageStreamCompleter(
    NetworkImage delegate,
    NetworkImage key,
    ImageDecoderCallback decode,
  ) {
    _executor.withResource(() async {
      final completer = delegate.loadImage(key, decode);
      final load = Completer<void>();
      completer.addListener(
        ImageStreamListener(
          (image, _) {
            setImage(image);
            if (!load.isCompleted) {
              load.complete();
            }
          },
          onChunk: reportImageChunkEvent,
          onError: (error, stackTrace) {
            reportError(exception: error, stack: stackTrace);
            if (!load.isCompleted) {
              load.complete();
            }
          },
        ),
      );
      await load.future;
    });
  }
}
```

</details>

**Comment by kevmoo** [comment]
Tracking issue filed on WebKit: https://bugs.webkit.org/show_bug.cgi?id=302801

We're actively validating mitigations.

**Comment by ronaldoandjames** [comment]
It seems that iOS Safari has very strict memory limits, and CanvasKit consumes more memory. When memory reaches a certain threshold, the tab is forcibly restarted.
        This issue has troubled us for many years, making our web almost unusable on iOS Safari. We had no choice but to downgrade Flutter to version 3.27.4 in order to use the HTML renderer. However, the HTML renderer has a “ghosting” issue when scrolling on the latest iOS Safari.
        We are now stuck on Flutter 3.27.4. 😭

**Comment by Garciconx** [comment]
Same issue here, guys, really sad :(

**Comment by kevmoo** [comment]
@Garciconx – this is the highest priority we have for issues. We have a SWE assigned. Please 👍 the issue. We'll stay in touch

**Comment by jeedy20** [comment]
A temporal fix will go a long way while we wait for the main fix

**Comment by Garciconx** [comment]
Any update?

**Comment by Bylinkk** [comment]
Hi, @kevmoo @yjbanov @mdebbar,
Thank you for keeping us updated on this problem.

**Comment by kevmoo** [comment]
I'm asking the engineers now. Thanks for your patience!

**Comment by harryterkelsen** [comment]
I'm trying to repro the problem with an app that loads many 20+ MB PNG images. However, I am using an iOS emulator, not a real device. I should have access to a real iOS device tomorrow and I'll try it on an iPhone.

In the meanwhile, is anyone who is hitting this bug able to reproduce the crash on an emulator? If so, can you please post the repro here? Thanks.

**Comment by Bylinkk** [comment]
@harryterkelsen , Here is the issue that was closed due to being a duplicate, with the complete analysis and the minimum reproducible code. #178524
[https://github.com/flutter/flutter/issues/178524]

**Comment by ronaldoandjames** [comment]
This issue is likely caused by convaskit consuming too much memory,
exceeding the limits set by iOS Safari.
I refactored our app using other native H5 technologies. While there will
still be some crashes, it will be much better.
BADRI Yassine ***@***.***>于2026年4月29日 周三下午11:37写道：

**Comment by harryterkelsen** [comment]
I got access to an iPhone today and was able to reproduce the crash. I believe the root cause is that Safari on iPhone crashes when decoding too many large images at once. I changed the engine-side code so that we throttle concurrent image decode jobs and the app no longer crashes on my iPhone. I also added instrumentation to our memory management code and observed that the total number of images was kept low, around 5-6 live images at a time, even though my app scrolls through 16 different large images. This gives me confidence that we are correctly disposing of and re-decoding large images over the app lifecycle.

I'll clean up my implementation of throttling the decode requests and I should have a PR out that fixes this issue by tomorrow.

I'll also double check that the repro in https://github.com/flutter/flutter/issues/178524 works with the fix. @Bylinkk , what do you mean when you say "Navigate between screens" in your reproduction app. Closing Safari and navigating to a different app?

**Comment by Bylinkk** [comment]
@harryterkelsen 
Thank you for your actions, this is great news!
What I mean by 'Navigate between screens' is that I've implemented a gesture detector with Navigator.pushNamed that returns to the same view. The navigator often doesn't crash on the first display, but it crashes when navigating between views multiple times. The images don't need to be large; a few kilobytes are enough to crash the browser.

**Comment by harryterkelsen** [comment]
@Bylinkk I'm not able to reproduce the crash on your example app on my iPhone (on either the latest main branch or stable 3.41.9)

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Image Crash Repro',
      initialRoute: '/',
      routes: {
        '/': (context) => const ImageListPage(),
      },
    );
  }
}

class ImageListPage extends StatelessWidget {
  const ImageListPage({super.key});

  final List<String> imageUrls = const [
    'https://picsum.photos/800/400?random=1',
    'https://picsum.photos/800/400?random=2',
    'https://picsum.photos/800/400?random=3',
    'https://picsum.photos/800/400?random=4',
    'https://picsum.photos/800/400?random=5',
    'https://picsum.photos/800/400?random=6',
    'https://picsum.photos/800/400?random=7',
    'https://picsum.photos/800/400?random=8',
    'https://picsum.photos/800/400?random=9',
    'https://picsum.photos/800/400?random=10',
    'https://picsum.photos/800/400?random=11',
    'https://picsum.photos/800/400?random=12',
    'https://picsum.photos/800/400?random=13',
    'https://picsum.photos/800/400?random=14',
];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Network Image Repro')),
      body: ListView.builder(
        // We multiply the list to ensure there is plenty of content to scroll
        itemCount: imageUrls.length * 4,
        itemBuilder: (context, index) {
          final url = imageUrls[index % imageUrls.length];
          return GestureDetector(
            onTap: () => Navigator.pushNamed(context, '/'),
            child: Card(
              margin: const EdgeInsets.all(12.0),
              child: Column(
                children: [
                  Padding(
                    padding: const EdgeInsets.all(12.0),
                    child: Text('Image #${index + 1}'),
                  ),
                  Image.network(
                    url,
                    fit: BoxFit.cover,
                    height: 400,
                    width: double.infinity,
                    loadingBuilder: (context, child, progress) {
                      if (progress == null) return child;
                      return Container(
                        height: 400,
                        color: Colors.grey[200],
                        child: const Center(child: CircularProgressIndicator()),
                      );
                    },
                  ),
                ],
              ),
            ),
          );
        },
      ),
    );
  }
}
```

I can get my iPhone to crash when I go to [img.bylinkk.com](https://img.bylinkk.com), however, that website is very slow and janky even on my Macbook Chrome. I suspect it has something to do with all of the iframe platform views and not the images. However, it's difficult to tell because the site is minified. If you can send the source code or rebuild the site with --profile I can try to diagnose the issue.

**Comment by harryterkelsen** [comment]
It will take a bit more time to implement the proper solution here. Just throttling the decodes would be detrimental to the performance of apps which have many small pictures. We don't want to decode the small pictures sequentially if the browser can handle decoding all of the small pictures in parallel. So, I played around with making throttling based on the total size of the images being decoded, but for <img> elements that are downloaded from a URL, we don't know the size until the browser starts decoding anyways. I'm working on a more robust fix and should have it out in a couple of days.

**Comment by jeedy20** [comment]
@harryterkelsen Keep it up. We desperately need this fix for work.

**Comment by jeedy20** [comment]
Hello @harryterkelsen has the issue been fixed?

**Comment by jeedy20** [comment]
If there is no possible fix, please bring back HTML rendering on the web. This is disrupting my work.

**Comment by ronaldoandjames** [comment]
I strongly agree.

Oliver Precious Chukwuemeka ***@***.***>于2026年6月17日
周三下午5:30写道：

**Comment by Garciconx** [comment]
Any update?

**Comment by harryterkelsen** [comment]
To give an update on this issue:

I previously addressed the iOS Safari crash by introducing an `ImageDecodingManager` that throttled simultaneous `HTMLImageElement.decode()` calls. This significantly reduced concurrent memory spikes and helped alleviate the OOM crashes.

Currently, I am in the process of landing a major refactor to the web engine's image decoding architecture (unifying the CanvasKit and Skwasm pipelines). As part of this refactor, we are completely removing the legacy HTML image element-based codec. Instead, for Safari and other browsers that do not support the native `ImageDecoder` API, we are now using `createImageBitmap` for static image decoding. 

`createImageBitmap` is a much more robust browser API and does not suffer from the same concurrency bugs and limitations as the old `HTMLImageElement.decode()` method. 

Once this unification refactor lands, we will monitor the performance and memory behavior of `createImageBitmap` on iOS Safari. If we find that throttling `createImageBitmap` calls is still beneficial for keeping peak memory usage low on highly resource-constrained iOS devices, I will re-introduce the `ImageDecodingManager` throttling in a follow-up PR.

**Comment by Bylinkk** [comment]
@harryterkelsen Thank you for your efforts to resolve this critical issue for our applications. Can we test these fixes in a beta version? It appears the fixes haven't been implemented in 3.46.0-0.2.pre, as the bug is still present. When do you plan to include them, at least in a beta release? Thanks in advance for your feedback.

**Comment by fobos** [comment]
@harryterkelsen The same for my team. We are still catching the bug.
Thanks for your efforts.

**Comment by fobos** [comment]
@harryterkelsen Hi, any updates? 
Thanks

**Comment by harryterkelsen** [comment]
The image decoding refactor has landed. I'm going to verify that decoding many large images is better with a real iOS device and report back

**Comment by harryterkelsen** [comment]
By the way, you can work around this issue pretty effectively by setting `webHtmlElementStrategy: WebHtmlElementStrategy.prefer` if you are using `Image.network`. Safari seems to heavily optimize `<img>` elements which make them much faster than creating a bitmap to upload to WebGL (the normal codepath on web). If you can, please try that out and let me know if it works for you.

**Comment by Bylinkk** [comment]
Thank you @harryterkelsen for your work on this! Do you know which Flutter version this patch will be included in?

**Comment by Bylinkk** [comment]
Thank you for the suggestion. I tested `webHtmlElementStrategy: WebHtmlElementStrategy.prefer`, and the issue is still present, but it occurs much less quickly than before.

When navigating rapidly between pages, Safari now crashes after dozens of very fast navigations, whereas previously it would only take a few navigations to crash the browser.

I also noticed that the issue takes significantly longer to occur when using smaller/lighter images.

So this workaround definitely improves the situation, but unfortunately it doesn't completely resolve the issue in my case.

**Comment by Bylinkk** [comment]
You can test it using the following link. Here is the video demonstrating the issue.
[https://devmode.scan.feadys.com](url)

https://github.com/user-attachments/assets/e47e835c-dcca-412d-9784-865dbd0ab1c4

**Comment by harryterkelsen** [comment]
@Bylinkk thanks for the repro. I'll investigate shortly
